### PR TITLE
Save extmark undo information when writing to undofile

### LINKED ISSUE #13048: [treesitter] errors with undofile
Cross-linking to neovim core project since this is related to https://github.com/neovim/neovim/blob/3c5141d2cf359d9ac70799bb987482d753263782/runtime/lua/vim/treesitter/query.lua

ref: https://github.com/nvim-treesitter/nvim-treesitter/issues/528

Ping @bfredl and @vigoux for 👀

- `nvim --version`: `NVIM v0.5.0-f7aeac726`
- `vim -u DEFAULTS` (version: ) behaves differently? yes; treesitter specific bug
- Operating system/version: `macOS Catalina 10.15.7`
- Terminal name/version: `alacritty 0.5.0 (a6681e3)`
- `$TERM`: `tmux-256color`

### Steps to reproduce
1. In vimrc: `set undofile`
2. Create a file with multiple-highlighted lines (like sample.bash)

```bash
export LDFLAGS="-L/usr/local/opt/llvm/lib"
export CPPFLAGS="-I/usr/local/opt/llvm/include"
export PATH="/usr/local/opt/llvm/bin:$PATH"
export SDKROOT=$(xcrun --sdk macosx --show-sdk-path)
```

3. edit the file, delete some lines, save and quit
4. hit `u` to undo changes; treesitter errors at this point and prints text into the editor area (although text "underneath" can still be highlighted/selected)

Screen Shot 2020-10-04 at 7 08 47 PM

### Actual behaviour

Treesitter errors and message is printed over textarea; the text "underneath" can still be edited/highlighted in visual mode

### Expected behaviour
* Treesitter should not error
* Treesitter errors should be printed into :messages
* Treesitter should be able to ingest and understand byte-changes from undofiles

### PR DESCRIPTION
**Pull request by chentoast**
Fixes #13048.

@bfredl  I keep getting a ton of detected leaks from Leak Sanitizer whenever using treesitter - most of these are from `nlua_regex` in executor.c. Is this an actual leak, or is this just because memory ownership is taken by lua and LSAN doesn't know that?

### CODE REVIEW

File: src/nvim/undo.c [hunk 1]
```diff
@@ -928,9 +939,129 @@ static u_header_T *unserialize_uhp(bufinfo_T *bi,
...
+      ILOG("Updating:");
+    if (!undo_write(bi, (uint8_t *)&(extup.data.savepos), sizeof(ExtmarkSavePos))) {
+      return false;
+    }
+  } else if (extup.type == kExtmarkSavePos) {
+      ILOG("Saving Position:");
+    if (!undo_write(bi, (uint8_t *)&(extup.data.savepos), sizeof(ExtmarkSavePos))) {
+      return false;
+    }
+  }
```
Comment (chentau) [inline src/nvim/undo.c:989]: Not sure which union element corresponds to `kExtmarkClear` and `kExtmarkUpdate`.

**Comment by bfredl** [comment]
@chentau not sure. I frequently get leaks with ASAN build locally but only a few objects. though I think we close lua state on exit (with ASAN build).

File: src/nvim/undo.c (hunk 1 above)
Comment (bfredl) [reply src/nvim/undo.c:988]: those are dead definitions, you can delete them.

**Comment by bfredl** [comment]
we should probably bump `UF_VERSION` . We might end up doing incompatible changes in 0.6 (extmark undo refactor), and then we can bump it again.

**Comment by chentau** [comment]
@bfredl  Ready for review :)

File: test/functional/api/extmark_spec.lua [hunk 2]
```diff
@@ -1423,6 +1423,48 @@ describe('API/extmarks', function()
...
+    -- reload undofile and undo changes
+    command("w")
+    command("set noundofile")
+    command("bw!")
+    command("e! testfile")
+
+    command("set undofile")
+
+    feed("u")
+    eq({"345", "hello world"}, curbufmeths.get_lines(0, -1, true))
```
Comment (bfredl) [inline test/functional/api/extmark_spec.lua:1460]: hmm, all of this behavior still works from before right? perhaps it is easier to test with `lua/buffer_update_spec.lua` in the `on_bytes` part (getting accurate byte-level updates).

Comment (chentau) [reply test/functional/api/extmark_spec.lua]: Makes sense. I've updated the tests

**Comment by chentau** [comment]
@bfredl Should I update the following head wiki page after this is merged?  I don’t see a way for this change to not break previous undofiles. Hopefully that won’t be too big of an issue.

**Comment by bfredl** [comment]
@chentau I think so. I think we consider undofile serialization of runtime state (for crash recovery etc), rather than long time storage. In case there is a mutiny we can add back compatibility of previous version (which shouldn't be too hard, but not worth it if no one uses it).

**Comment by chentau** [comment]
@bfredl Is there anything else I need to do to help get this merged?

File: src/nvim/undo.c [hunk 3]
```diff
@@ -928,9 +944,90 @@ static u_header_T *unserialize_uhp(bufinfo_T *bi,
...
+    extup = unserialize_extmark(bi, &error, file_name);
+    if (error) {
+      kv_destroy(uhp->uh_extmark);
+      xfree(extup);
+      return NULL;
+    }
+    kv_push(uhp->uh_extmark, *extup);
+    xfree(extup);
+  }
+
```
Comment (bfredl) [inline src/nvim/undo.c:962]: copy the `if (c != UF_ENTRY_END_MAGIC)` block here.

**Comment by bfredl** [comment]
I think this is pretty much done. But we need to communicate clearly with master-followers so it doesn't come as a surprise :]

**Comment by akinsho** [comment]
I'm looking forward to this change but just curious _what_ a user will have to be warned of? will the old undo files error, will they need to be all deleted or will they continue to work but still reproduce the original bug and to get the benefits of this PR they will need to removed so new compatible ones can be created?

**Comment by chentau** [comment]
@bfredl Should be ready for merge, unless you spot anything else.

**Comment by chentau** [comment]
@akinsho As of right now, the behavior will be the latter; since this format of the undofile has changed, any undofiles before this pr will be outdated and won't work. If re-creating all undofiles proves to be too inconvenient for people, then I'll submit another PR implementing backwards compatibility, although then the old undofiles will still errror out with treesitter.

**Comment by akinsho** [comment]
@chentau this particular bug is easily the most annoying thing I face on a daily basis whilst editing so really don't mind trading off against losing undofiles they are a nice to have vs. error free highlighting 😄 , was more just curious about what to expect.

**Comment by bfredl** [comment]
Merged. New versions of neovim will not read undofiles written by nvim before this merge (there will be an error message about incopmatible version). Nvim 0.4 (or an master up to bda12927be84f0e0e05e885f9acb40a7ac2c5524 ) can be used to recover older undofiles, so if you worried about unsaved changes lurking around in undofiles it would make sense to keep such version around somewhere to recover them.

This is a necessary change to keep tree-sitter and plugins dependent on byte-level buffer change events fully working with undo states from a undofile. If there is a clear demand we might implement reading of the old format. Such recovered buffers will not be be fully functional with plugins relying on buffer updates or tree-sitter, however.

**Comment by jamessan** [comment]
> If there is a clear demand we might implement reading of the old format.

It's a versioned file format.  Why wouldn't we keep the ability to read the old version?

> Such recovered buffers will not be be fully functional with plugins relying on buffer updates or tree-sitter, however.

There's not a logical way to transform the old version to the new version?

**Comment by bfredl** [comment]
> It's a versioned file format. Why wouldn't we keep the ability to read the old version?

IIUC undofiles is intended to be more of a crash recovery or session suspend-restore thing than permanent storage of history. for instance: if the file contents doesn't match the current state in the undofile exactly the entire undofile already becomes unusable. So I think the version is more for avoid reading incompatible file by mistake, rather than version multiple files in a permanent storage.

> There's not a logical way to transform the old version to the new version?

Not really. undo headers in vim operate on entire changed lines only, while we are adding byte-level change resolution. we could, in principle, construct faked byte edits that replace the entire line region, though this will cause performance regressions with tree-sitter and unexpected behaviours with plugins using byte resolution updates (which includes extmarks).

**Comment by lervag** [comment]
I'm sorry about the silly question, but: It seems this change creates a definite incompatibility with Vim. I know there is no goal of staying compatible, but I've actually been able to keep compatibility until this moment. Do I understand correctly, or am I missing something simple that can establish compatibility again?

**Comment by chentau** [comment]
@lervag As far as I'm aware, there's no easy way to establish compatibility going forward. This is because we are serializing additional byte level information about edits, and Vim has no analogue of this. I guess the reason I made this PR is because without this extra information, undofiles will not work with any plugins that leverage the `on_bytes` mechanism, which include plugins like tree-sitter and basically anything that uses extmark highlighting/virtual text. 

That being said, if you have an idea for how to implement compatibility going forward, I'd be happy to submit another PR, since I  agree that the breaking of compatibility is unfortunate.

**Comment by bfredl** [comment]
@lervag if you mean _mutual_ compatibility i e open the same file back and forward both in vim and neovim, then that will require also _vim_ storing detailed byte tracking in undo headers. The thing is tho is that the current format is not really what we want (it works for tree-sitter but is not detailed enough for LSP for instance) but for 0.6 I hope to implement more fully-featured undo tracking and then we should seriously discuss with vim-dev about a new shared format.

**Comment by lervag** [comment]
Thanks @chentau and @bfredl, I appreciate your responses and clarifications. My main "concern" here is that, as a plugin author, there will be slightly more of a hazzle to keep my plugins compatible with both Vim and neovim (since I will now probably decouple my Vim and neovim configs).

As a happy user of neovim, I strongly respect and acknowledge all of the very good work you and all the other devs are doing. And I don't mind that your development of neovim will require me both to learn new stuff (e.g., I will probably have to learn Lua at some time :p) and to make some choices (such as this: Vim vs neovim).

**Comment by Bluesboy** [comment]
My old undo-files aren't working anymore doesn't it breaks compatibility?

**Comment by bfredl** [comment]
@Bluesboy https://github.com/neovim/neovim/wiki/Following-HEAD

**Comment by blueyed** [comment]
I think it might be worth reading the old format, and writing/converting it to the new format.  This should get enabled explicitly though however.

As for me I'm using the following for now, i.e. a separate dir for nvim-0.5, which results in missing undo information now however..!

```vim
if has('persistent_undo')
  if has('nvim-0.5')
    " New format in https://github.com/neovim/neovim/pull/13973 (f42e932,
    " 2021-04-13).
    let &undodir = s:vimsharedir . '/undo2'
  else
    let &undodir = s:vimsharedir . '/undo'
  endif
  …
endif

**Comment by MRAAGH** [comment]
Undofile incompatibility should probably be mentioned in [:h vim-differences](https://github.com/neovim/neovim/blob/997a9c879215bc01a928de3e762955878314ec6a/runtime/doc/vim_diff.txt)

**Comment by davidchisnall** [comment]
> IIUC undofiles is intended to be more of a crash recovery or session suspend-restore thing than permanent storage of history

I strongly disagree with this.  The fact that I have undo history for files in source trees that I have not yet committed, that is preserved across reboots and isn't lost when I go away from the tree for a few months, is by far my favourite vim feature: it respects Raskin's first law.

This was my first interaction with neovim: It could not read any of the history that I have in my undofiles directory (~/.cache/vim/undo/ for me) and so lost state.  The state was recoverable by using vim, but because the filename is not changed I have to modify my vimrc to configure vim and neovim to use different undo locations.

This could have avoided violating the principle of least astonishment by changing the filename.  It could have avoided sharp data loss by reading the old format but writing only the new format.  If the new format had a different name then this could have been transparent: the old file would have the undo history up to the point neovim 0.5 was used and so you could undo right back to the start, but new changes would be visible only with the new editor.

**Comment by bfredl** [comment]
> it respects Raskin's first law.

No idea what this law says, but vim has already broken it twice in the past. :)

> This was my first interaction with neovim: It could not read any of the history that I have in my undofiles directory (~/.cache/vim/undo/ for me) and so lost state. [...] This could have avoided violating the principle of least astonishment by changing the filename.

Hmm, but the filename is different already, neovim uses `~/.cache/nvim/undo/`. The issue you mention would occur when switching from _neovim_ 0.4 to 0.5 but not with first usage of neovim.

> . It could have avoided sharp data loss by reading the old format but writing only the new format. 

It shouldn't overwrite an unreadable file tho, unless I have completely misread the code. Seems when you say "sharp data loss" and "lost state"  you actually mean unreadable by new version but still recoverable by using old version? These conditions differ quite a lot in severity so best to not confuse them :)

Reading old files in a fallback mode have already been suggested several times. As said before it is doable in principle but the open question is what happen when using such a buffer with byte state dependent plugins (such as all tree-sitter plugins).

**Comment by davidchisnall** [comment]
> No idea what this law says, but vim has already broken it twice in the past. :)

A program must not harm a user's data or through inaction allow a user's data to come to harm.

> Hmm, but the filename is different already, neovim uses ~/.cache/nvim/undo/. 

Neovim reads `$VIMINIT`, which for me is set to `source ${XDG_CONFIG_HOME}/vimrc`.  It therefore reads my vimrc file, which sets the undo location.  I had to modify my `vimrc` with an `if has("nvim")` block to set these locations.

> The issue you mention would occur when switching from neovim 0.4 to 0.5 but not with first usage of neovim.

Or, as in my case, with first usage of neovim with a vimrc that sets `undodir`.

> It shouldn't overwrite an unreadable file tho, unless I have completely misread the code. 

It doesn't, but because it reads the vim-specific config file from the vim-specified location it then writes a file in that location with the name that vim expects, so files that I first edited in neovim ended up with undo files that vim couldn't read and vice versa.  Both programs complained.

I've heard great things about neovim, but this caused enough frustration in my first 10 minutes of trying to use it that I ended up uninstalling it and reverting back to vim.

**Comment by MRAAGH** [comment]
> I ended up uninstalling it and reverting back to vim.

Personally I've elected to stick with neovim 0.4.4, it's a great program and Vim-compatible. Just my 2 cents.

**Comment by EdmundsEcho** [comment]
@blueyed Thank you for posting the temporary fix.  Your suggestion that the v0.6 convert the old undo format to the new is a good one.  Clearly the effort required is likely non-trivial because the desired outcome (non-breaking change), is clearly the dominant choice.  The suggestion to fix this, is it something to expect in the iteration or so?  

The error 824 suggests the cause is that the undo was written using a *newer* version... not as is the case where the new version (0.6) cannot read the old version.

\- E

**Comment by dkiyatkin** [comment]
When I want to see old undo history, before lose it, I use this docker-command with nvim 0.4.4. Maybe someone find it usefull:
```
docker run -it -v `pwd`:/mnt/volume --workdir=/mnt/volume anatolelucet/neovim:0.4.4 nvim --cmd "set undodir=." --cmd "set undofile" my_file_with_old_undo
```

**Comment by Arnie97** [comment]
I didn't have enough patience to read `undo.c` and build a real `.un~` parser.
However I wrote a crappy script that simulates the changes in this patch and just works in my (insufficient) tests: [ufdump](https://github.com/Arnie97/ufbump/blob/master/ufbump)

**Comment by EricDuminil** [comment]
@Arnie97 : Thanks a lot for your script! It seems to have worked fine, and upgraded my precious undofiles from 0.4.4 to 0.8.2!

**Pull request merged**